src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Variables.cs

```diff
@@ -0,0 +1,373 @@
+                }
+                slot = -1;
+                return false;
+            }
+
+            internal int Add(VariableIdentifier identifier)
+            {
+                if (Count > IndexMask)
+                {
+                    return -1;
+                }
+                if (GetSlotDepth(identifier.ContainingSlot) >= MaxSlotDepth)
+                {
+                    return -1;
+                }
+                var variables = GetVariablesForVariable(identifier);
+                return variables.AddInternal(identifier);
+            }
+
+            private int AddInternal(VariableIdentifier identifier)
+            {
+                int index = Count;
+                _variableSlot.Add(identifier, index);
+                while (index >= _variableBySlot.Count)
+                {
+                    _variableBySlot.Add(default);
+                }
+                _variableBySlot[index] = identifier;
+                return ConstructSlot(Id, index);
+            }
+
+            internal bool TryGetType(Symbol symbol, out TypeWithAnnotations type)
+            {
+                var variables = GetVariablesContainingVariable(symbol);
+                return variables._variableTypes.TryGetValue(symbol, out type);
+            }
+
+            internal void SetType(Symbol symbol, TypeWithAnnotations type)
+            {
+                var variables = GetVariablesContainingVariable(symbol);
+                Debug.Assert((object)variables == this);
+                variables._variableTypes[symbol] = type;
+            }
+
+            internal VariableIdentifier this[int slot]
+            {
+                get
+                {
+                    (int id, int index) = DeconstructSlot(slot);
+                    var variables = GetVariablesForId(id);
+                    return variables!._variableBySlot[index];
+                }
+            }
+
+            internal int Count => _variableBySlot.Count;
+
+            internal void GetMembers(ArrayBuilder<(VariableIdentifier, int)> builder, int containingSlot)
+            {
+                (int id, int index) = DeconstructSlot(containingSlot);
+                var variables = GetVariablesForId(id)!;
+                var variableBySlot = variables._variableBySlot;
+                for (index++; index < variableBySlot.Count; index++)
+                {
+                    var variable = variableBySlot[index];
+                    if (variable.ContainingSlot == containingSlot)
+                    {
+                        builder.Add((variable, ConstructSlot(id, index)));
+                    }
+                }
+            }
+
+            private Variables GetVariablesForVariable(VariableIdentifier identifier)
+            {
+                int containingSlot = identifier.ContainingSlot;
+                if (containingSlot > 0)
+                {
+                    return GetVariablesForId(DeconstructSlot(containingSlot).Id)!;
```

Comment:
>return GetVariablesForId(DeconstructSlot(containingSlot).Id)!; [](start = 20, length = 62)

It looks like this is just an alternative way to locate where container (containingSlot) belongs without traversing containing symbols. If that is the case, let's assert that through symbols we end up with the same result.

Reply:
`containingSlot` represents the receiver rather than the containing symbol, and the receiver is not available here.

---
In reply to: [561460142](https://github.com/dotnet/roslyn/pull/50417#discussion_r561460142) [](ancestors = 561460142)

Reply:
> containingSlot represents the receiver rather than the containing symbol, and the receiver is not available here.

But we can traverse to the top-most receiver and that one has a symbol. What am I missing?

---
In reply to: [562109896](https://github.com/dotnet/roslyn/pull/50417#discussion_r562109896) [](ancestors = 562109896,561460142)

Reply:
Added an assert to `Variables.Add()` that the ids of the child matches the id of the containing slot.

---
In reply to: [562118922](https://github.com/dotnet/roslyn/pull/50417#discussion_r562118922) [](ancestors = 562118922,562109896,561460142)